app_dart/lib/src/request_handling/swarming_authentication.dart

```diff
@@ -0,0 +1,135 @@
+  }) : super(
+          _config,
+          clientContextProvider: clientContextProvider,
+          httpClientProvider: httpClientProvider,
+          loggingProvider: loggingProvider,
+        );
+
+  /// The Cocoon config, guaranteed to be non-null.
+  final Config _config;
+
+  /// Provides the App Engine client context as part of the
+  /// [AuthenticatedContext].
+  ///
+  /// This is guaranteed to be non-null.
+  final ClientContextProvider clientContextProvider;
+
+  /// Provides the logger.
+  ///
+  /// This is guaranteed to be non-null.
+  final LoggingProvider loggingProvider;
+
+  static const String kAgentIdHeader = 'Agent-ID';
+
+  /// Name of the header that LUCI requests will put their service account token.
+  static const String kSwarmingTokenHeader = 'Service-Account-Token';
+
+  /// Authenticates the specified [request] and returns the associated
+  /// [AuthenticatedContext].
+  ///
+  /// See the class documentation on [AuthenticationProvider] for a discussion
+  /// of the different types of authentication that are accepted.
+  ///
+  /// This will throw an [Unauthenticated] exception if the request is
+  /// unauthenticated.
+  @override
+  Future<AuthenticatedContext> authenticate(HttpRequest request) async {
+    final String agentId = request.headers.value(kAgentIdHeader);
+    final String swarmingToken = request.headers.value(kSwarmingTokenHeader);
+
+    final ClientContext clientContext = clientContextProvider();
+    final Logging log = loggingProvider();
+
+    if (agentId != null) {
+      // Authenticate as an agent. Note that it could simultaneously be cron
+      // and agent, or Google account and agent.
+      final Key agentKey = _config.db.emptyKey.append(Agent, id: agentId);
+      final Agent agent = await _config.db.lookupValue<Agent>(agentKey, orElse: () {
+        throw Unauthenticated('Invalid agent: $agentId');
+      });
+
+      if (!clientContext.isDevelopmentEnvironment) {
+        final String agentAuthToken = request.headers.value('Agent-Auth-Token');
+        if (agentAuthToken == null) {
+          throw const Unauthenticated('Missing required HTTP header: Agent-Auth-Token');
+        }
+        if (!_compareHashAndPassword(agent.authToken, agentAuthToken)) {
+          throw Unauthenticated('Invalid agent: $agentId');
+        }
+      }
+
+      return AuthenticatedContext(agent: agent, clientContext: clientContext);
+    } else if (swarmingToken != null) {
+      return await authenticateIdToken(swarmingToken, clientContext: clientContext, log: log);
+    }
+
+    throw const Unauthenticated('Request rejected due to not from LUCI or Cocoon agent');
+  }
+
+  // This method is expensive (run time of ~1,500ms!). If the server starts
+  // handling any meaningful API traffic, we should move request processing
+  // to dedicated isolates in a pool.
+  static bool _compareHashAndPassword(List<int> serverAuthTokenHash, String clientAuthToken) {
```

Comment:
move to a common script so both `SwarmingAuthenticationProvider` and `Authentication` can share?

Reply:
Done. Made public in `AuthenticationProvider`, with plan to remove after devicelab luci migration is complete.